### ISSUE TITLE
Add additional options for collider rotation / position settings on the components

### ISSUE THREAD

**Issue by arionsh2** [Yellow-Dog-Man/Resonite-Issues#2775]
### Is your feature request related to a problem? Please describe.

Colliders are frustrating to include on avatars. Positioning them to contour correctly to the avatar often requires complex offsetting that I'd like to have auto update if I ever make tweaks. This is adding load as I often support these colliders with flux to get the calculations I'd like.

### Describe the solution you'd like

I'd like colliders updated with the following:

1. Add a "rotation offset" to each collider. This is most critical on the cylinder colliders, as not every mesh lines up nicely with the forward vector of the slot.
2. Add a "center offset on" dropdown. This offset would take global position data and find the midpoint of the two slots identified based on the following dropdown. The current "position offset" would still be present, and would offset from this new center point. Presets would be:
   - "Grandparent" (collider is centered in global space between the grandparent slot and the slot the component is on)
   - "Parent" (ditto, but only center between the parent and this slot)
   - "Custom" (unlocks 2 slot inputs for arbitrary center calculations)
   - "None" (current behavior where the collider starts centered on the current slot it is in). 
3. Add a "scale adjust" to each collider. This is most useful on the sphere, to achieve more irregular but rounded shapes. This would be like if the collider was on a dedicated slot and that slot's scale was being adjusted.

### Describe alternatives you've considered

Currently, to get all this functionality, I'm creating a cloud of slots that take in the "start point" / "end point" slots and do these calculations on the fly via flux. It's given me good coverage on the avatars I use, but I imagine packing it into the components themselves would make it more efficient and convenient.

### Additional Context

If the 3 additions needed prioritization, the "center at" option is what I think would be most immediately useful. Centering the collider so that it covers the range between a parent slot and its child often best matches the mesh visual. The other 2 are more "nice to have" ideas.

### Requesters

Arion Silverhoof / arionsh

**Comment by Frooxius** [comment]
You can already achieve most of these by simply adding a child Slot, adding collider that and transforming that. It's a proper way to handle this scenario and lets you transform the collider in whatever way you want with existing mechanisms and components to transforms slots - you could request more slot transform components to help there.

We're very unlikely to add these to colliders specifically for a few reasons:
1) It's already possible to achieve this behavior through existing mechanism (see above)
2) Adding this will add very significant engineering complexity to the collider/physics system and duplicate a lot of work that's already done for Slots
3) It would have no major befit to do so - in fact, this might end up having a global negative impact when those options aren't in use, because the system still has to deal with a lot of the overhead
4) These are fairly arbitrary requirements. Generally we follow "decomposition" approach, where each component and system has its own set of responsibilities and you combine them to get more complex behaviors. If we added these to colliders, there could be lots of other behaviors that could be added, resulting in large bloat.

> Add a "scale adjust" to each collider. This is most useful on the sphere, to achieve more irregular but rounded shapes. This would be like if the collider was on a dedicated slot and that slot's scale was being adjusted.

I do not quite understand this. You can scale colliders by scaling the slot they're on, but that will never change the shape of the collider. Sphere will still be sphere, even if non-uniformly scaled, because physics systems do not support non-uniform scales.

**Comment by arionsh2** [comment]
I'm gonna respond to these points out of order, but I totally understand where you're coming from.

>     4. These are fairly arbitrary requirements. Generally we follow "decomposition" approach, where each component and system has its own set of responsibilities and you combine them to get more complex behaviors. If we added these to colliders, there could be lots of other behaviors that could be added, resulting in large bloat.

This makes perfect sense, but colliders seem to exist in a weird nexus of "mesh object whose meaning depends on another mesh object". I've rarely found the need to center a collider on a single point in space (floors and walls being notable exceptions). Rather, I usually need to stretch a collider in response to any two points in space moving around, particularly on avatars. A collider centered on my elbow isn't terribly useful, but one that stretches from between my wrist and my elbow actually matches my physical shape. I'd like the component to have the tools it needs to be smart enough to do that without extra flux / slot management. Especially if the relative distance between the two points might modify (a button that gives me really long arms for a few minutes or whatever).

>     2. Adding this will add very significant engineering complexity to the collider/physics system and duplicate a lot of work that's already done for Slots

This is probably a really good reason not to do the enhancement request, though it is unfortunate. I personally treat colliders like special case slots, and need to exercise most of the controls a slot has to make them act the way I want them to. The rotation / scale adjust settings would be useful to pack more of those slot characteristics into the component itself, since I so frequently have to just put it on a dedicated slot anyway. My logic is that if the workflow is getting repeated a lot, it would "probably" be more efficient to have backend code ready to handle it automatically? I could also just have a really unorthodox workflow. I dunno, it seemed worth filing the request about and seeing what others thought.

> I do not quite understand this. You can scale colliders by scaling the slot they're on, but that will never change the shape of the collider. Sphere will still be sphere, even if non-uniformly scaled, because physics systems do not support non-uniform scales.

Huh... I thought if I scaled a slot with a sphere collider to like [1,2,1], it would distend the sphere to match the slot dimensions. It doesn't do that at all. Totally disregard that, then. Not sure where I got the mistaken idea from.

>     1. It's already possible to achieve this behavior through existing mechanism (see above)
>     3. It would have no major befit to do so - in fact, this might end up having a global negative impact when those options aren't in use, because the system still has to deal with a lot of the overhead

The major benefit would be that the collider component would have all the settings / controls needed to treat it as the mesh object it is right there on the component, no need for an extra slot. Also, when the collider registers a hit, it would be able to live directly on the armature for the avatar, and return the exact slot on the armature that was hit.

So, in summary, I'd agree that the direction of the request **is** to enable the collider component to act as a special case slot, because I'm conceptualizing tying it to bones on the armature. There are absolutely workarounds to keep from requiring this, I just anecdotally seem to use those workarounds most of the times I'm wanting to position a collider "correctly" for my needs. As I said above, since the workflow seemed to keep getting repeated, it seemed prudent to ask if it could be packaged together better.

Either way, thanks for considering and responding :)
